pkg/operator/cluster/creds_controller.go

```diff
@@ -0,0 +1,176 @@
+which also has the apache 2.0 license.
+*/
+
+package cluster
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"k8s.io/api/core/v1"
+
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/rook/rook/pkg/clusterd"
+)
+
+var (
+	// controllerResyncPeriod potentially can be part of rook Operator configuration.
+	// in seconds
+	controllerResyncPeriod time.Duration = 120
+	// controllerRetry is number of retries to process an event
+	controllerRetry int = 5
+)
+
+func NewCredsController(ns string, context *clusterd.Context) *PVCController {
+	return &PVCController{
+		clusterNamespace: ns,
+		clusterContext:   context,
+		queue:            workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+}
+
+// PVCController implements Kube controller that monitors rook PVC events.
+// The controller ensures all kube namespaces have a set of credentials
+// to manage lifeCycle of ceph primitives.
+type PVCController struct {
+	queue            workqueue.RateLimitingInterface
+	indexer          cache.Indexer
+	informer         cache.Controller
+	clusterContext   *clusterd.Context
+	clusterNamespace string
+}
+
+func (c *PVCController) pvcHandler(key interface{}) error {
+	obj, _, err := c.indexer.GetByKey(key.(string))
+	if err != nil {
+		return err
+	}
+
+	pvc := obj.(*v1.PersistentVolumeClaim)
+
+	if !strings.Contains(*pvc.Spec.StorageClassName, "rook") {
+		return nil
+	}
+
+	cephUser := &cephUser{c.clusterContext, c.clusterNamespace}
+	if err := cephUser.create(); err != nil {
+		return err
+	}
+
+	if err = cephUser.setKubeSecret(pvc.ObjectMeta.Namespace); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *PVCController) run(stopCh chan struct{}) {
+	eventList := cache.NewListWatchFromClient(
+		c.clusterContext.Clientset.Core().RESTClient(),
+		"persistentvolumeclaims",
+		v1.NamespaceAll,
+		fields.Everything(),
+	)
+
+	c.indexer, c.informer = cache.NewIndexerInformer(
+		eventList,
+		&v1.PersistentVolumeClaim{},
+		controllerResyncPeriod*time.Second,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.onAdd,
+			UpdateFunc: c.onUpdate,
+			DeleteFunc: c.onDelete,
+		},
+		cache.Indexers{},
+	)
+
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+	logger.Info("Starting PVC controller (managing ceph creds).")
+
+	go c.informer.Run(stopCh)
+
+	// Wait for all involved caches to be synced, before processing items from the queue is started
+	if !cache.WaitForCacheSync(stopCh, c.informer.HasSynced) {
+		runtime.HandleError(fmt.Errorf("Timed out waiting for caches to sync"))
+		return
+	}
+
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	<-stopCh
+	logger.Info("Stopping PVC (ceph creds) controller.")
+}
+
+func (c *PVCController) runWorker() {
+	for c.processNextItem() {
+	}
+}
+
+func (c *PVCController) processNextItem() bool {
+	key, quit := c.queue.Get()
```

Comment:
Interesting how you did this. Why not use the client-go informer object. Here is a example https://godoc.org/k8s.io/client-go/tools/cache#NewInformer

 With the informer object, you dont need to queue things as it is already has a store. Perhaps this would make it cleaner